src/rgw/rgw_auth.h

```diff
@@ -122,14 +122,21 @@ class IdentityApplier : public Identity {
  *  E. execute-commit - commit the modifications from point C. */
 class Completer {
 public:
-  typedef std::unique_ptr<Completer> cmplptr_t;
+  /* It's expected that Completers would tend to implement many interfaces
+   * and be used not only in req_state::auth::completer. Ref counting their
+   * instances woild be helpful. */
+  typedef std::shared_ptr<Completer> cmplptr_t;
```

Comment:
It's [passed](https://github.com/ceph/ceph/pull/14885/files#diff-e0ff328397db5c0f824c3f867d8eed0fR966) to the front-end subsystem as the AWSv4 completers are also filters over IO.